File: food_delivery_backend/restaurant/serializers/restaurant.py
from rest_framework import serializers

from food.models import DishCategory
from restaurant.models import (
    Restaurant,
    RestaurantCategory,
    RestaurantLike,
)

# from restaurant.serializers import (
#     BasicInfoSerializer, RepresentativeSerializer,
#     DetailInfoSerializer, MenuDeliverySerializer
# )

from restaurant.serializers.basic_info import BasicInfoSerializer
from restaurant.serializers.detail_info import DetailInfoSerializer
from restaurant.serializers.menu_delivery import MenuDeliverySerializer
from restaurant.serializers.representative_info import RepresentativeInfoSerializer
from restaurant.serializers.payment_info import PaymentInfoSerializer
# from order.serializers.owned_promotion import (
#     RestaurantPromotionSerializer
# )
from food.serializers import DishCategorySerializer, DishSerializer
from review.serializers import RestaurantReviewSerializer
from order.serializers.promotion import RestaurantPromotionSerializer

from utils.serializers import CustomRelatedModelSerializer
from utils.pagination import CustomPagination
from  utils.function import get_related_url_3


class RestaurantSerializer(CustomRelatedModelSerializer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.many_related_serializer_class = {
            'dishes': DishSerializer
        }

    dishes = serializers.SerializerMethodField()
    distance_from_user = serializers.SerializerMethodField()
    is_liked = serializers.SerializerMethodField()

    def get_dishes(self, obj):
        requrest = self.context.get('request', None)
        query_params = requrest.query_params if requrest else None
        category = query_params.get('category', None)

        queryset = obj.dishes.all() \
            if category is None \
            else obj.dishes.filter(category__name=category)
        
        paginator = CustomPagination(page_size_query_param='dish_page_size', page_query_param='dish_page')
        request = self.context.get('request')
        page = paginator.paginate_queryset(queryset, request)

        return DishSerializer(page, many=True, context=self.context).data

    def get_distance_from_user(self, obj):
        request = self.context.get('request')
        if request and hasattr(request, 'user') and hasattr(request.user, 'locations'):
            user_location = request.user.locations.filter(is_selected=True).first()
            if user_location:
                return obj.basic_info.get_distance_from_user(user_location)
        return None

    def get_is_liked(self, obj):
        return obj.is_liked(
            request=self.context.get('request'),
        )

    class Meta:
        model = Restaurant
        fields = [
            'id', 
            'basic_info', 
            'distance_from_user', 
            'dishes', 
            'rating', 
            'total_reviews', 
            'total_likes',
            'avg_price',
            'is_certified',
            'is_liked',
        ]
        depth = 1
    
class DetailRestaurantSerializer(CustomRelatedModelSerializer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.one_related_serializer_class = {
            'basic_info': BasicInfoSerializer,
            'detail_info': DetailInfoSerializer,
            'payment_info': PaymentInfoSerializer,
            'representative_info': RepresentativeInfoSerializer,
            'menu_delivery': MenuDeliverySerializer,
        }
        self.many_related_serializer_class = {
            # 'categories': {
            #     'serializer': DishCategorySerializer,
            #     'context': {'detail': True}
            # },
            
            # 'promotions': PromotionSerializer,
            # 'owned_promotions': RestaurantPromotionSerializer,
            # 'user_reviews': RestaurantReviewSerializer
        }

    distance_from_user = serializers.SerializerMethodField()
    stats = serializers.SerializerMethodField()
    is_liked = serializers.SerializerMethodField()
    categories = serializers.SerializerMethodField()


    def get_distance_from_user(self, obj):
        request = self.context.get('request')
        if request and hasattr(request, 'user') and hasattr(request.user, 'locations'):
            user_location = request.user.locations.filter(is_selected=True).first()
            if user_location and hasattr(obj, 'basic_info'):
                return obj.basic_info.get_distance_from_user(user_location)
        return None
    
    def get_stats(self, obj):
        request = self.context.get('request')
        return get_related_url_3(
            request=request,
            obj=obj,
            action='stats',
            detail=True
        )

    def get_is_liked(self, obj):
        return obj.is_liked(
            request=self.context.get('request'),
        )

    def get_categories(self, obj):
        queryset = [x.category for x in obj.restaurant_categories.filter(is_disabled=False)]
        return DishCategorySerializer(queryset, many=True, context=self.context).data

    class Meta:
        model = Restaurant
        exclude = [
            'user', 
            # 'categories',
        ]


class CreateRestaurantSerializer(serializers.ModelSerializer):
    class Meta:
        model = Restaurant
        fields = ['user',]
        read_only_fields = ['id',]

    def to_representation(self, instance):
        return DetailRestaurantSerializer(instance, context=self.context).data
    
class UpdateRestaurantSerializer(serializers.ModelSerializer):
    categories = serializers.ListField(
        child=serializers.UUIDField(),
        required=False,
    )
    disabled_categories = serializers.ListField(
        child=serializers.UUIDField(),
        required=False,
    )

    class Meta:
        model = Restaurant
        fields = [
            'categories',
            'disabled_categories',
        ]

    def update(self, instance, validated_data):
        restaurant = instance
        categories = validated_data.pop('categories', [])
        disabled_categories = validated_data.pop('disabled_categories', [])
        updated_categories = []
        updated_disabled_categories = []

        print('categories', categories, pretty=True)
        print('disabled_categories', disabled_categories, pretty=True)

        for _category in categories:
            _instance, _created = RestaurantCategory.objects.get_or_create(
                restaurant_id=restaurant.id,
                category_id=_category
            )
            _instance.is_disabled = False
            _instance.save()
            updated_categories.append(_instance)
        print(updated_categories, pretty=True)

        for _category in disabled_categories:
            _instance, _created = RestaurantCategory.objects.get_or_create(
                restaurant_id=restaurant.id,
                category_id=_category
            )
            _instance.is_disabled = True
            _instance.save()
            updated_disabled_categories.append(_instance)
        print(updated_disabled_categories, pretty=True)

        return super().update(instance, validated_data)

    def to_representation(self, instance):
        return DetailRestaurantSerializer(instance, context=self.context).data
    

class RestaurantCategorySerializer(serializers.ModelSerializer):
    category = serializers.SerializerMethodField(read_only=True)

    def get_category(self, obj):
        # context = self.context.update({
        #     'detail': False
        # })
        return DishCategorySerializer(obj.category, context=self.context).data
    
    class Meta:
        model = RestaurantCategory
        fields = '__all__'


class CreateUpdateRestaurantCategorySerializer(serializers.ModelSerializer):
    class Meta:
        model = RestaurantCategory
        fields = [
            'id',
            'restaurant',
            'category',
            'created_at',
            'is_disabled',
        ]
        read_only_fields = ['id', 'created_at',]
    

class RestaurantLikeSerializer(serializers.ModelSerializer):
    class Meta:
        model = RestaurantLike
        fields = "__all__"
        read_only_fields = ['id', 'created_at',]

File: food-delivery-app/lib/features/authentication/models/restaurant/restaurant.dart
import 'package:flutter/foundation.dart';
import 'package:food_delivery_app/data/services/reflect.dart';
import 'package:food_delivery_app/features/authentication/models/account/user.dart';
import 'package:food_delivery_app/features/authentication/models/restaurant/basic_info.dart';
import 'package:food_delivery_app/features/authentication/models/restaurant/detail_info.dart';
import 'package:food_delivery_app/features/authentication/models/restaurant/menu_delivery.dart';
import 'package:food_delivery_app/features/authentication/models/restaurant/payment_info.dart';
import 'package:food_delivery_app/features/authentication/models/restaurant/representative_info.dart';
import 'package:food_delivery_app/features/user/food/models/food/category.dart';
import 'package:food_delivery_app/features/user/food/models/food/dish.dart';
import 'package:food_delivery_app/features/user/food/models/review/review.dart';
import 'package:food_delivery_app/utils/helpers/helper_functions.dart';

@reflector
@jsonSerializable
class Restaurant {
  final String? id;
  final String? user;
  final RestaurantBasicInfo? basicInfo;
  final RestaurantRepresentativeInfo? representativeInfo;
  final RestaurantDetailInfo? detailInfo;
  final RestaurantMenuDelivery? menuDelivery;
  final RestaurantPaymentInfo? paymentInfo;
  final List<DishCategory> categories;
  final dynamic dishes;
  final double rating;
  final double avgPrice;
  final int? totalReviews;
  final String? restaurantReviews;
  final Map<String, dynamic> ratingCounts;
  final bool isCertified;
  final String? stats;
  final String? deliveries;
  final bool isLiked;
  int totalLikes;
  final String? restaurantCategories;
  final String? promotions;
  // final List<RestaurantPromotion>? promotions;

  Restaurant({
    this.id,
    this.user,
    this.basicInfo,
    this.detailInfo,
    this.paymentInfo,
    this.representativeInfo,
    this.menuDelivery,
    this.dishes = const [],
    this.categories = const [],
    this.rating = 0,
    this.avgPrice = 0,
    this.totalReviews,
    this.restaurantReviews,
    this.ratingCounts = const {},
    this.isCertified = false,
    this.stats,
    this.deliveries,
    this.isLiked = false,
    this.totalLikes = 0,
    this.restaurantCategories,
    this.promotions,
    // required this.promotions,
  });

  Restaurant.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        user = json['user'],
        basicInfo = json['basic_info'] != null ? RestaurantBasicInfo.fromJson(json['basic_info']) : null,
        representativeInfo = json['representative_info'] != null ? RestaurantRepresentativeInfo.fromJson(json['representative_info']) : null,
        detailInfo = json['detail_info'] != null ? RestaurantDetailInfo.fromJson(json['detail_info']) : null,
        menuDelivery = json['menu_delivery'] != null ? RestaurantMenuDelivery.fromJson(json['menu_delivery']) : null,
        paymentInfo = json['payment_info'] != null ? RestaurantPaymentInfo.fromJson(json['payment_info']) : null,
        categories = json['categories'] != null ? (json['categories'] as List).map((instance) => DishCategory.fromJson(instance)).toList() : [],
        dishes = json['dishes'] != null ?
            json['dishes'] is String ? json['dishes'] :
              (json['dishes'] as List).map((instance) => Dish.fromJson(instance)).toList() : [],
        rating = THelperFunction.formatDouble(json['rating']),
        avgPrice = THelperFunction.formatDouble(json['avg_price']),
        totalReviews = json['total_reviews'],
        restaurantReviews = json['restaurant_reviews'],
        ratingCounts = json['rating_counts'] ?? {},
        isCertified = json['is_certified'] ?? false,
        stats = json['stats'],
        deliveries = json['deliveries'],
        isLiked = json['is_liked'] ?? false,
        totalLikes = json['total_likes'] ?? 0,
        restaurantCategories = json['restaurant_categories'],
        promotions = json['promotions']
        ;


  Map<String, dynamic> toJson({ bool patch = false}) {
    Map<String, dynamic> data = {
      'id': id,
      'user': user,
      'basic_info': basicInfo?.toJson(),
      'representative_info': representativeInfo?.toJson(),
      'detail_info': detailInfo?.toJson(),
      'menu_delivery': menuDelivery?.toJson(),
    };

    if(patch) {
      data.removeWhere((key, value) => value == null);
    }

    return data;
  }

  String get formatTotalReviews {
    return THelperFunction.formatNumber(totalReviews ?? 0);
  }

  @override
  String toString() {
    return THelperFunction.formatToString(this);
  }
}

@jsonSerializable
@reflector
class RestaurantLike {
  final String? id;
  final String? restaurant;
  final String? user;
  final DateTime? createdAt;

  RestaurantLike({
    this.id,
    this.restaurant,
    this.user,
    this.createdAt,
  });

  RestaurantLike.fromJson(Map<String, dynamic> json)
    : id = json['id'],
      restaurant = json['restaurant'],
      user = json['user'],
      createdAt = json['createdAt'] != null ? DateTime.parse(json['created_at']) : null
  ;

  Map<String, dynamic> toJson({ bool patch = false}) {
    Map<String, dynamic> data = {
      'user': user,
      'restaurant': restaurant,
    };

    if(patch) {
      data.removeWhere((key, value) => value == null);
    }

    return data;
  }

  @override
  String toString() {
    return THelperFunction.formatToString(this);
  }
}


@jsonSerializable
@reflector
class RestaurantCategory {
  final String? id;
  final dynamic restaurant;
  final dynamic category;
  final DateTime? createdAt;
  final bool isDisabled;

  RestaurantCategory({
    this.id,
    this.restaurant,
    this.category,
    this.createdAt,
    this.isDisabled = false,
  });

  RestaurantCategory.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        restaurant = json['restaurant'] is List || json['restaurant'] is String || json['restaurant'] == null
            ? json['restaurant']
            : DishCategory.fromJson(json['restaurant']),
        category = json['category'] is List || json['category'] is String || json['category'] == null
            ? json['category']
            : DishCategory.fromJson(json['category']),
        createdAt = json['createdAt'] != null ? DateTime.parse(json['created_at']) : null,
        isDisabled = json['is_disabled'] ?? false
  ;

  Map<String, dynamic> toJson({ bool patch = false}) {
    Map<String, dynamic> data = {
      'category': category is DishCategory ? category?.id : category,
      'restaurant': restaurant is Restaurant ? restaurant?.id : restaurant,
      'is_disabled': isDisabled
    };

    if(patch) {
      data.removeWhere((key, value) => value == null);
    }

    return data;
  }

  @override
  String toString() {
    return THelperFunction.formatToString(this);
  }
}
